summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -0,0 +1,132 @@
+package com.twitter.summingbird.storm.collector
+
+import com.twitter.algebird.Semigroup
+import com.twitter.summingbird.online.executor.KeyValueShards
+import com.twitter.summingbird.online.option.SummerBuilder
+import backtype.storm.spout.{ ISpoutOutputCollector, SpoutOutputCollector }
+import com.twitter.algebird.util.summer.{ AsyncSummer, Incrementor }
+import com.twitter.util.{ Await, Duration, Future, Time }
+import scala.collection.mutable.{ Map => MMap, MutableList => MList }
+import scala.collection.{ Map => CMap }
+import scala.collection.JavaConverters._
+import java.util.{ List => JList }
+
+/**
+ *
+ * AggregatorOutputCollector is a wrapper around the SpoutOutputCollector.
+ * AsyncSummer is used to aggregate the tuples.
+ * Different streams have separate aggregators and caches.
+ */
+class AggregatorOutputCollector[K, V: Semigroup](
+    in: SpoutOutputCollector,
+    summerBuilder: SummerBuilder,
+    summerShards: KeyValueShards,
+    flushExecTimeCounter: Incrementor,
+    executeTimeCounter: Incrementor) extends SpoutOutputCollector(in) {
+
+  // An individual summer is created for each stream of data. This map keeps track of the stream and its corresponding summer.
+  private val spoutCaches = MMap[String, AsyncSummer[(K, V), Map[K, V]]]()
+
+  // As the crushDown happens at a stream level. We have a mapping from stream to the messageIds.
+  // The messageIds are further aggregated to the level of summerShards.
+  // The Map keeps track of aggregated tuples' messageIds as a list.
+  private val streamMessageIdTracker = MMap[String, MMap[Int, MList[Object]]]()
+
+  /**
+   * This method is invoked from the nextTuple() of the spout.
+   * This is triggered with tick frequency of the spout.
+   */
+  def timerFlush(): Unit = {
+
+    val startTime = Time.now
+    spoutCaches.foreach {
+      case (stream, cache) =>
+        val tupsOut = cache.tick.map { convertToSummerInputFormat(_) }
+        emitData(tupsOut, stream)
+    }
+    flushExecTimeCounter.incrBy(Time.now.inMillis - startTime.inMillis)
+  }
+
+  private def convertToSummerInputFormat(flushedCache: CMap[_, _]): CMap[Int, CMap[_, _]] =
+    flushedCache.groupBy { case (k, _) => summerShards.summerIdFor(k) }
+
+  /*
+    The method is invoked to handle the flushed cache caused by
+    exceeding the memoryLimit, which is called within add method.
+   */
+  private def emitData(cache: Future[TraversableOnce[(Int, CMap[_, _])]], streamId: String): JList[Integer] = {
+    val startTime = Time.now
+    val flushedTups = Await.result(cache)
```

Comment:
I haven't seen users configure AsyncSummer implementation much at all. I've seen only one job so far that specifies AsyncSummer implementation explicitly. The other case is Async flag in tsar, which would work bad in this case. There are very few sync implementations of AsyncSummer any way. There are more chances of users ending up in a suboptimal state. IMO, better to have restricted functionality than something we know has pitfalls.